### ISSUE TITLE
Feature request: Add UI for testing full alerting expressions over time

### ISSUE THREAD

**Issue by qed-** [prometheus/prometheus#1154]
It would be great if there was a way to generate time series (from recording rules and synthetic alert timeseries).

This would effectively allow backtesting of alerts and recording rules over past data.

**Comment by juliusv** [comment]
@qed- For my better understanding, how would this behave differently from just evaluating the alert expression as a graph to see at which historical points it generates output time series? That's how I currently test alert expressions over history.

For better alert testing usability, I'd like us to add a UI input field though that allows you to test an entire alerting rule (`ALERT <alertname> IF <query expression> ...`), not just the query expression that the rule contains.

**Comment by qed-** [comment]
If I understand right it's possible to graph the alert expression but not any FOR and WITH clauses.

Ideally you could click on an alert and see exactly how it would have been evaluated in the past, complete with labels.

**Comment by brian-brazil** [comment]
Labels are currently static, so `WITH` doesn't add much. Even when that's more powerful what you mainly care about is if alerts will fire (so you can calibrate the threshold) rather than the exact labels involved.

**Comment by juliusv** [comment]
Yes, the thing about `FOR` is true though. That should be covered by my suggestion about having a UI way to test full alerting expressions (and being able to click on existing configured ones to do the same would be a great addition to that).

**Comment by qed-** [comment]
Yes, I mentioned labels as being the way to tell if the alert is firing or pending.

**Comment by juliusv** [comment]
So in my suggestion, the following would happen:

There's a UI textarea that lets you paste in an entire alerting expression. Alternatively, you can click on an already configured alerting rule to take you to the pre-filled-in version of that.

When pressing a button, you get a graph over time where you basically see labeled time series for any firing alert vector elements at the times when they are firing. If nothing is firing over the entire history, you would just get an empty graph. This is where Prometheus's alerting model differes fundamentally from one like Bosun's, where an alerting rule always has a single scalar output value vs. a whole vector of possible values (or none).

As an additional bonus, one could display a graph of just the *count* of firing alert elements over time or something like that.

Would that address your feature request?

**Comment by juliusv** [comment]
(and that would take into account the `FOR` stanzas etc. of course)

**Comment by qed-** [comment]
That would be perfect.

**Comment by juliusv** [comment]
@qed- Great, I've adjusted the title of this issue to reflect this more clearly. I can't say yet when this is going to happen, but I think it would be a great thing to have in general.

**Comment by brian-brazil** [comment]
I think in addition we need something to handle the `for` clauses. A summary of how many alerts fired for how long would be useful for tuning that.

**Comment by mattbostock** [comment]
Related: https://github.com/prometheus/prometheus/pull/1369

**Comment by kadaan** [comment]
I have resurrected @codesome's https://github.com/prometheus/prometheus/pull/4277, made some improvements, and it work well for my use cases.  I have included it in my https://github.com/kadaan/promutil.  

<img width="1514" alt="2022-09-01 at 13 41 58@2x" src="https://user-images.githubusercontent.com/1904898/188008592-0712b708-0c41-4d11-af62-3a1aaade337d.png">

**Comment by codesome** [comment]
@kadaan thats awesome. I wonder if you would be willing to contribute it to Prometheus upstream? I think it is a useful feature to have. The API work and the UI could be done separately.

**Comment by kadaan** [comment]
@codesome Sure, I would love to contribute it back. It might take me a bit as it has to fit in with my day job.  One thing to be aware of is that while it can run against any Prometheus server, I operate against a local one that has data pulled from out prod server using the other tools listed below. This is important to me as the potential for negative impact on server performance is high, especially if the number of alert rules tested is large or the time range under test is huge.

For my workflow Promutil has two other commands that are super useful that I don’t think have existing parallels in promtool: migrate and generate. 

Migrate runs remote api queries over a time range (days to weeks) against a Prometheus server and writes the resultant data to a local TSDB directory. 

Generate uses an expression language to generate metrics over a time range and writes them to a local TSDB folder. 

The way I end up using these for alert or recording rule testing is to run a migrate first from our prod cluster to my laptop. Then I run a generate against my laptop for any data I need  doesn’t yet exist in prod. Once that is done I can iterate as much as I want, against my local TSDB, testing alert rules over large time ranges or backfilling recording rules and verifying they work as expected. This is nice because the potential impact to a prod cluster is zero once you migrate the data to a local TSDB. Promutil ships with a docker-compose file which will run Prometheus and Grafana locally, so I can also easily get dashboard testing out of the deal. 

Lastly, there is choice made during alert rule evaluation which I was forced into for performance concerns. Evaluating alert labels and annotations can be extremely costly as they can spawn a high number of additional Prometheus queries. In this case, we only care about that information for those points of the alert lifecycle that are transitions between states: inactive -> pending, pending -> active, and active -> resolved. As such, I evaluate all the alerts and record the timestamps of these state transitions. Then I am able to evaluate the alerts again at these state transition timestamps, but this time resolve the labels and annotations. This greatly reducing the performance impact of additional queries from labels and annotations.

**Comment by Tehada** [comment]
@kadaan , hey, how's the progress? I am urgently in need of backtesting of alerts to tune them for my infra and I am willing to put some effort into implementation!

**Comment by krajorama** [comment]
Hello from the bug scrub!

@juliusv do you think this would make sense to add this to the new UI? Do we want to add such features to the new UI?

It seems like there were attempts at working on this and the use case stands, so keeping open.
